### PR TITLE
Lua via sol2

### PR DESCRIPTION
**Pull request by olanti-p**
### Summary
SUMMARY: Infrastructure "Bring back Lua"

### Purpose of change
Experimental PR to bring back Lua modding.

According to DDA issue tracker and numerous discussions on Discord:
1. Having some scripting in game would be great. Not only it'd allow more interesting mods, it may also attract new modders with Lua background.
2. Porting Effect-On-Condition system from DDA 0.F+ will take a lot of effort due to sheer volume of changes and bugfixes. It's also a Domain-Specific Language, with all the issues that come from it (we'll have to manually expand language functionality, there are no libraries or tutorials, modders will have to learn it).
3. Bringing back Lua in Delta Fork by undoing [its removal](https://github.com/CleverRaven/Cataclysm-DDA/pull/28572) was a pain due to amount of code differences and time passed since then. Also, it resolves none of the issues of the old implementation.
4. The old Lua was "a PITA to maintain" due to multiple reasons:
    * Lua stack is hard
    * Lua build was fragile
    * You had to run a Lua script to generate bindings before compilation
    * You had to manually define function signatures and classes in a lua table and keep them up to date
5. Lua had little use in core, essentially turning it into deadweight and catalyzing bitrot. Part of it was from Lua being originally implemented as debugging API, part of it - from frequent build issues, part of it - from removal plans discouraging modders from sinking effort into new mods.

I can't magically make point 5 disappear, but the rest of the issues should be solvable.

As such, this PR uses completely different approach from the one that was used in DDA before Lua got axed.

The old Lua integration used Lua stack directly and then relied on scripts to generate C++ bindings file from a predefined Lua list that mirrored C++ definitions.

This version relies on a Lua C++ wrapper called [sol2](https://github.com/ThePhD/sol2) to hide Lua stack management from the developer, so creating new bindings can be done by adding a few lines of human-readable C++. It still has to be done manually, but at least sol2 is able to automatically figure out types of objects being bound, so it's not much different from our de-/serialization code.

I'm not planning on re-implementing entire Lua functionality right off the bat, but this PR (if it goes in) should give a solid foundation to build on.

### Describe the solution
The idea is to make Lua integration:
1. build and work consistently on all builds/platforms
2. require little to no additional build setup from dev side
3. be simple enough to be understood by people who haven't spent days and weeks studying Lua stack and virtual machine
4. be optional
5. be as nice as possible when it comes to using from Lua side

As such, the current solution is:
1. Add a folder in `src` named `lua` and put Lua 5.3.6 source code there. Lua is extremely fast to compile (it takes less than 10 seconds to compile on my machine when using single thread), so having it be built alongside other game source files and then be linked in one step mitigates a whole slew of issues (Lua version availability between platforms, Lua version mismatch between build machine and host machine, system Lua installation weirdness on some platforms)
2. Add a folder in `src` named `sol` and put sol2 v3.3.0 single-header source code there. Sol2 is a mature, feature-rich and performant C++ Lua wrapper that should cover many of our potential needs.
3. Add some automated tests (and Github Actions workflows) to help deal with compile- and run-time problems when it comes to building with Lua enabled.
4. Run an example Lua script on game start.
5. (Implemented in #2233) Refactor and streamline game package loading. Had to get rid of `core` non-mod (that's been causing trouble before btw, see #2106 ), and also of per-world mods (it's an obscure and almost unused mechanic with no access from UI).
6. On load stage, run Lua scripts provided by mods. The loading scheme is inspired by Factorio, see `Data loading` part in `doc/LUA_SUPPORT.md` for details.
7. Add debug Lua console. It can run simple commands, entire scripts and also shows `log_` output from Lua functions
8. Add Lua hot-reload functionality by allowing to reload some of the mods' scripts on keypress & from menu
9. Add a bunch of bindings and hooks available for use by Lua mods
10. Implement a thin wrapper for registering usertypes dubbed `luna`. The purpose of this wrapper is to provide types registered with `sol2` with additional runtime information such as type name as registered in Lua (`sol` only keeps track of type name as it's specified in C++, in cases where there is mismatch it becomes impossible to know Lua type from userdata) for (de-)serialization of Lua tables with userdata. It also emits additional information on members registered for the usertype, such as field names and types, method names and their arguments and return types.
11. Add a `--lua-doc` runtime switch that makes the game generate documentation for registered types based on information collected by `luna`. This way, we can have some rudimentary automatic docs generation for bindings that does not rely on some external tools.  
12. Add a small `Smart House Remotes` mod that uses Lua to open/close curtains and garage doors in a house
13. Add a small `SaveLoad Lua Test` mod that uses hooks to preserve its Lua state throughout save/load cycle
14. In LUA=0 builds, show warning on mods that use Lua and disallow loading worlds that use Lua mods

### Describe alternatives you've considered
Main alternatives and their disadvantages I listed in introduction.

Alternatives regarding current solution:
1. Use custom `clang-tidy` plugin to generate bindings.
    Unfortunately I don't know where to even begin writing one. I have little experience with `clang-tidy`, and the current tidy-with-plugins installation workflow could be somewhat hard for new contributors, see [comment](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/2216#issuecomment-1332732496).
    Upd: turns out clang's syntax parsing can be used via a standalone library, so it should be possible to create a standalone bindings generator and just have the developers/contributors download precompiled releases.
    Upd2: same as with `clang-tidy`, it requires a lot of knowledge regarding clang codegen

2. Use some other Lua bindings library.
    I remember using [LuaBridge](https://github.com/vinniefalco/LuaBridge) a few years back, and sol2 is definitely nicer to work with. Maybe they've improved over the years, I didn't check. Also after some googling, sol2 seems to be the go to solution for C++ <-> Lua bindings in _current year_ due to its overall performance, flexibility and C++iness.

### Testing
Added PR workflows to build and test linux g++ and macos clang++ builds.

The tests include basic stuff like creating lua state, loading Lua libraries, bindings some definitions and running code.
* `lua_class_members` - Binds `point` class and its members, various constructors and methods (including math operators), then runs a Lua script that uses these bindings
* `lua_global_functions` - Binds a couple global functions and creature class family, then runs a Lua script that calls the global function and verifies that bound `get_character_name` accepts class `avatar` which is derived from `Character` from Lua side while original C++ function expects `Character`.
* `lua_called_from_cpp` - Runs a Lua script that creates and stores a function, then executes that function from C++ side and validates the results.
* `lua_runtime_error` - Causes runtime error on Lua side and checks that we receive appropriate exception on C++ side.
* `lua_called_error_on_lua_side` - Calls Lua's `error()` on Lua side and checks that we receive appropriate exception on C++ side.
* `lua_called_error_on_cpp_side` - Calls Lua's `error()` on C++ side and checks that we receive appropriate exception on C++ side.
* `lua_called_cpp_func_throws` - Throws exception from C++ function called by Lua and checks that we receive appropriate exception on C++ side.
* `lua_table_serde*` - Checks correctness of saving & loading a Lua table to/from JSON
* `id_conversion*` - Checks id conversion functions (raw pointer <-> string_id <-> int_id)

### TODO list (this PR)
- [x] ~~Investigate possibility of writing automated bindings generator using `libclang`~~ - Way to complex for me to tackle. Also, not sure if this is viable given how diverse the codebase is; it's way easier to write the bindings semi-manually. 
- [x] Expose more stuff to Lua
- [x] Set up some Lua hooks
- [x] Run Lua scripts provided by mods, figure out mod interop (I'm leaning towards copying Factorio's Lua loading scheme)
- [x] Build a tiny but full-fledged mod showcasing Lua functionality
- [x] Make Lua available in Windows builds
- [x] Make Lua available in CMake builds
- [x] Make Lua available in Android builds
- [x] Look into semi-automatic Lua doc generation
- [x] Update docs
- [x] ~~Gather feedback~~ Feedback would have to be gathered after merging
- [x] Figure out localization for Lua mods (~~may just end up making wrappers to gettext calls, `xgettext` supports Lua~~ `extract_json_strings.py` can now parse Lua source and extract strings from relevant `*gettext` calls)
- [x] Lua table (de-)serialization to/from JSON
- [x] Handle usertypes and edge cases in table (de-)serialization (see TODOs in code)
- [x] ~~Look for a way to abstract away differences between `int_id<T>` vs `string_id<T>` vs `const T*` on Lua side~~ - It seems `sol` does not allow customizing how `const T*` is treated for certain types. Added manual conversion functions.
- [x] ~~Workaround the sol2 issue where binding `string_id<T>` requires `T` to implement operators `<` and `==`~~ - It would require either removing comparison operators from `string_id<T>` and `int_id<T>`, removing `operator*` and `operator->` from them or patching `sol2`. I don't like all of these variants, meanwhile implementing comparison operators for the types is way simpler and doesn't kill ergonomics. Added a macro to make implementing them easier, it will have to do for now.
- [x] Look for a way to split up `catalua_bindings.cpp` as it will get progressively slow to compile
- [x] Introduce Lua API versioning ~~and maybe game and mod versioning, to avoid potential mismatches here~~ Ended up adding Lua API versioning in form of single integer, will have to see how it works out.

### TODO list (long-term)
- [ ] Maybe a clang-tidy plugin to generate documentation from bindings? This would be easier than the current `luna` wrapper.
- [ ] Look for a nice way to bind typesafe coords
- [ ] Look for more options in mod interop
- [ ] Come up with naming convention for Lua side
- [ ] Expose even more stuff to Lua
- [ ] Add even more hooks
- [ ] Lua API testing infrastructure
- [ ] Write tests for API
- [ ] Make Lua mandatory and rewrite "content" parts of the codebase to use Lua (e.g. item use, player activities, examine funcs, more intricate parts of mapgen)

### CODE REVIEW

**Comment by Coolthulhu** [comment]
> Maybe a clang-tify plugin to generate documentation from bindings?

I thought about using `clang-tidy` to generate bindings.
A good binding generator will certainly be needed to maintain Lua in long term. One of the reasons it died in DDA was the necessity to bind them semi-manually.

**Comment by olanti-p** [comment]
> using `clang-tidy` to generate bindings

Tidy would be great, but I've given it some thought over the day, and I'm having trouble imagining how to make it work easily.
* The plugin itself - it'll probably end up quite complex, so whoever writes it will have to spend a _lot_ of time learning llvm internals or have some preexisting knowledge of the topic
* To run the plugin we have to use a patched version of clang-tidy, which according to [the doc](https://github.com/cataclysmbnteam/Cataclysm-BN/blob/d0df8ee6fe1e854f3372d159c157b94515891527/doc/DEVELOPER_TOOLING.md#custom-clang-tidy-plugin) has to be compiled from source for everything except Ubuntu Xenial
* There seems to be no instructions for running clang-tidy with cata plugins on macos
* Would the Lua build of the game require generating bindings? If so, does that mean macos won't have Lua builds?
* Only CMake solution has configuration for clang-tidy with cata plugins. What about Visual Studio and Make? Do we ditch them and migrate to CMake?
* If someone PRs a C++ change without testing how it'd affect bindings (e.g. someone on Windows who couldn't set up a custom build of llvm and has given up halfway - Lua is optional, so why bother if it kinda works), how do we test and fix it?

I'm not sure what else could be used as an alternative though.

One option I've considered is to have a PR job that runs a single custom tidy plugin that's basically a tweaked version of DDA's [tidy serialize check](https://github.com/CleverRaven/Cataclysm-DDA/blob/0b1daba70f8780f38d69157351e65d021ca40fcc/tools/clang-tidy-plugin/SerializeCheck.cpp) that instead of checking whether member is serialized would check whether members and methods are bound to Lua. It would still be half-manual, but at least CI would point exact place where manual intervention is required.  Another job could scrape the C++ files for bindings (e.g. have it match every `sol::state::new_usertype<T>(...)` and `sol::usertype<T>::operator[](...)` invocations), and then build documentation as human-readable webpage and as machine-friendly text file that'll be diffed against previous version to see what API has been added/removed/changed.

**Comment by olanti-p** [comment]
Putting this on hold for now, clang codegen needs to be investigated + with #2250 we would have item manipulation API that's more consistent overall and more friendly to Lua in particular. Related: #2265

**Comment by olanti-p** [comment]
Okay, I've "investigated" clang codegen and discovered that thing is _way_ over my head.

It basically has the same biggest issue as implementing generator via tidy plugin: someone would need to be familiar with libtooling, or spend some considerable time learning it, to be able to write such tool and extend its functionality. This is not something I would be able to pull off on top of everything else I already do for BN, not without quitting my day job.

**Comment by joveeater** [comment]
Yeah libtooling is incredibly complex but that's what you get when you try to parse c++. I want to make something that lets us just give a class name and then runs a lambda on each method in that class and that should hopefully cover a good chunk of both of our needs. We'll probably also want something to help us with types but that will need more thinking through. 
I'm gonna go ahead with this once #2250 is done but I won't use it for actors until after doing more game object stuff with creatures and vehicles.

File: src/game.cpp [hunk 1]
```diff
@@ -1512,6 +1513,17 @@ bool game::do_turn()
...
+    cata::run_on_every_second( *DynamicDataLoader::get_instance().lua );
+    if( calendar::once_every( 1_minutes ) ) {
+        cata::run_on_every_minute( *DynamicDataLoader::get_instance().lua );
+    }
+    if( calendar::once_every( 1_hours ) ) {
+        cata::run_on_every_hour( *DynamicDataLoader::get_instance().lua );
+    }
+    if( calendar::once_every( 1_days ) ) {
+        cata::run_on_every_day( *DynamicDataLoader::get_instance().lua );
+    }
```
[the remark this answers is not in the document]
Comment (olanti-p) [reply src/game.cpp:1525]: It's more of a temporary solution to provide people with some recurrent hooks to play with.
I'd like to add some dedicated hook registering API, where hooks would have unique identifiers so mods could alter them, remove, or even alter and remove hooks of other mods, and that would include stuff like "run this hook every `time_duration`", or "run this hook once after `time_duration` has passed", or "run this hook every `time_duration`, but only as long as you're on z-level -10 and within 8 omts of a lab finale"

**Comment by olanti-p** [comment]
Ok, I'm calling it "good enough".

**Comment by Coolthulhu** [comment]
HUEG

Will start reviewing

File: Makefile [hunk 2]
```diff
@@ -928,6 +944,14 @@ else
 	@$(RC) $(RFLAGS) $< -o $@
 endif
 
+$(ODIRLUA)/%.o: $(LUA_SRC_DIR)/%.c
+ifeq ($(VERBOSE), 1)
+	$(CXX) -xc -std=c11 -c $< -o $@
+else
+	@echo $(@F)
+	@$(CXX) -xc -std=c11 -c $< -o $@
+endif
```
Comment (Coolthulhu) [inline Makefile:953]: Explicit c++11? Is it needed for Sol2?

Comment (olanti-p) [reply Makefile:953]: Oh, it's for Lua itself. It is written in C, and I just used C11 as a moderately recent C standard. I think the original Lua makefile used C99.

Sol2 itself is header-only, so it's not in the build files.

**Comment by Coolthulhu** [comment]
The way I understand it, right now, it still requires manually binding C++ structs to Lua ones. Is that right?
It's OK for now, but it is not maintainable in the long term, so we will certainly need a better solution in the future.

**Comment by olanti-p** [comment]
Yes.
It is semi-automatic, as in sol handles actual Lua stack operations, argument validation and overload/inheritance resolution, and the luna wrapper handles doc generation for function arguments/return values and struct member types, but in the end you still need to manually define that "`class Character` has function `i_rem` that must be bound under name `remove_item`".

**Comment by Coolthulhu** [comment]
I can't get the smart remote to work. It always prints "no signal", regardless of whether I spawn it or find one spawned by mapgen.

**Comment by olanti-p** [comment]
It's a bit of a hidden mechanic, but the house it's linked to must have a grid with charged battery. It'll say "no signal" if there's no charge or the character is too far away.

**Comment by Coolthulhu** [comment]
Alright, got it working.
Does it have to be both mapgen-spawned and powered? If so, that's pretty much debug-only and the mod should probably be marked as obsolete.

**Review by Coolthulhu — APPROVED** [review]

**Comment by olanti-p** [comment]
Marked it as obsolete.

File: doc/LUA_SUPPORT.md [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+game.cata_internal          For internal game purposes, please don't use this
+game.hooks.<hook_id>        Hooks exposed to Lua scripts, will be called on corresponding events
+game.iuse.<iuse_id>         Item use functions that will be recognized by the item factory and called on item use
+
+### Game Bindings
+The game exposes various functions, constants and types to Lua.
+Functions and constants are organized into "libraries" for organizational purposes.
+Types are available globally, and may have member functions and fields.
+
+To see the full list of funcitons, constants and types, run the game with `--lua-doc` command line argument.
```
Comment (ZhilkinSerg) [inline doc/LUA_SUPPORT.md:179]:
```suggestion
To see the full list of functions, constants and types, run the game with `--lua-doc` command line argument.
```

**Comment by scarf005** [comment]
https://luals.github.io/wiki/annotations/

generating definition files would be great, as their syntax doesn't differ much from lua docs generator

**Pull request merged**